src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs

```diff
@@ -272,10 +277,10 @@ string EncodeUrl (Uri url)
 				if(Logger.LogNet)
 					Logger.Log (LogLevel.Info, LOG_APP, " cancelled");
 
-				httpConnection.Disconnect ();
+				await DisconnectAsync (httpConnection);
 				cancellationToken.ThrowIfCancellationRequested ();
 			}
-			cancellationToken.Register (httpConnection.Disconnect);
+			cancellationToken.Register (async () => await DisconnectAsync (httpConnection));
```

Comment:
@alanmcgovern: what's wrong is that the [`CancellationToken.Register()`](https://msdn.microsoft.com/en-us/library/dd321635(v=vs.110).aspx) docs don't make explicitly clear on which thread the `Action` will be invoked.

In this case, we _do not_ want `httpConnection.Disconnect()` to be invoked from the UI thread, and since the docs don't say what thread will invoke the callback, the `async () => ...` is used to "move" it to a different thread.

Or such is the hope.